### ISSUE TITLE
make twistd pluggable, nerf mktap

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1922]
|<img alt="tjs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tjs reported|
|-|-|
|Trac ID|trac#1922|
|Type|enhancement|
|Created|2006-07-14 00:48:46Z|

twistd <subcommand> should work, and use <subcommand> in a way very similar to how mktap uses it, the difference being that twistd will run the service immediately instead of saving it to a file.

Suggested usage: 
```
$ twistd -n web --path .
<startup messages>
```

```
$ twistd web --path .
# forks and daemonizes.
```

<details><summary>Searchable metadata</summary>

```
trac-id__1922 1922
type__enhancement enhancement
reporter__tjs tjs
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1152838126000000 1152838126000000
changetime__1156714330000000 1156714330000000
version__None None
owner__ 
cc__jerub cc__spiv cc__itamarst cc__exarkun cc__ralphm cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

Writing out .tac files is perhaps worth doing, but I don't think it should be a suggested deployment mechanism. See [#11551](https://github.com/twisted/twisted/issues/11551) for my proposal.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

minus one trillion to writing out tac files.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_4"></a>|@radix set owner to @radix<br>@radix set status to `assigned`|
|-|-|

Indeed minus one trillion to writing out tac files. I'm changing the description to leave that bit out. I'm also removing the stuff about not writing a tistd.log or .pid. I've started working on a branch (twistd-plugins-1922) to implement the remaining features.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_5"></a>|@radix removed owner<br>@radix set status to `new`|
|-|-|

Ok I think maybe this is ready for review in twistd-plugins-1922.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set owner to @ralphm<br>@ralphm set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @radix<br>@exarkun set status to `new`|
|-|-|

This needs to add some new unit tests for the new behavior.

I suggest at least these two tests:

  * One which grabs subCommands directly and asserts that subcommands for a set of test plugins are provided.  The test plugins can live in a test directory, and you can parameterize the package to look plugins up in on the ServerOptions class.
  * One which calls twisted.scripts.twistd.getApplication directly with various configurations and asserts something about the returned Application

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Oops.  I completely missed the added test_twistd.py.

Revised comments:

  * use file instead of open
  * don't delete the tap file in tearDown
  * I'd still like to see a test that actually results in getPlugins being called, but the tests here are pretty decent, so I guess that's optional.
  * add test-case-name to scripts/twistd.py

Merge when done

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_10"></a>|@radix set status to `closed`|
|-|-|

(In [17926]) Merge twistd-plugins-1922

Author: radix
Reviewer: exarkun
Fixes [#1922](https://github.com/twisted/twisted/issues/1922)

The first step towards a mktapless world.

"Oh, ye so fiercely tended,
  Ye little seeds of hate!
I bent above your growing
  Early and noon and late,
Yet are ye drooped and pitiful,--
  I cannot rear ye straight!"

twistd now can accept subcommands equivalent to mktap's;
it will look up the IServiceMaker plugin for the given subcommand
and use it in precisely the same way that mktap does, except that it
immediately applies the service to an Application and starts it.

That is to say,

"twistd web --path ."

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_11"></a>|@radix commented|
|-|-|

(In [17959]) Revert r17926 - doesn't work on Windows (Fuck Yes I Said It)
Refs [#1922](https://github.com/twisted/twisted/issues/1922)

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_12"></a>|@radix commented|
|-|-|

(In [17962]) Reviving dead branch twistd-plugins-1922
Refs [#1922](https://github.com/twisted/twisted/issues/1922)

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_14"></a>|@radix commented|
|-|-|

(In [17965]) *really* revive twistd-plugins-1922
Refs [#1922](https://github.com/twisted/twisted/issues/1922)

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_16"></a>|@radix commented|
|-|-|

I think this is ready for review again. It probably works on Windows now, as I've now got unit tests that test the code there, but I haven't tested it interactively.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_17"></a>|@radix commented|
|-|-|

ahem, and the branch is twistd-plugins-1922-2.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `new`|
|-|-|

ApplicationRunner.createOrGetApplication should comment its usage of loadedPlugins, describing how iterating over subCommands initializes and populates it.

UnixApplicationRunner and its methods should have docstrings.

Any reason not to make setupEnvironment, startApplication, and startLogging (private) methods the Runner classes?

Nothing tests that preApplication and postApplication are ever called or are called in the right order with respect to makeService.

Itamar thinks there should be a subprocess test for twistd (but I think it'd be terrible).

Seems to basically work on Windows.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_19"></a>|@radix set status to `assigned`|
|-|-|

The private functions in _twistd_unix: I'm not changing their API because certain third parties have decided to rely on them. When a more concerted effort is made to refactor twistd to have a decent internal design, then it will be more worthwhile to break those APIs.

Thanks for the review, I'm working on the other issues.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_20"></a>|@radix set owner to @exarkun<br>@radix set status to `new`|
|-|-|

Ok I've done everything you required in your review.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_21"></a>|@radix set owner to @radix<br>@radix set status to `assigned`|
|-|-|

exarkun gives +1:

```
[17:01] exarkun radix: you don't know how to spell privileges
...
[17:03] exarkun fix that and then I guess merge
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_22"></a>|@radix set status to `closed`|
|-|-|

(In [17992]) Merge twistd-plugins-1922-3

Author: radix
Reviewer: exarkun
Fixes [#1922](https://github.com/twisted/twisted/issues/1922)

The first step towards a mktapless world: take three.

"Oh, ye so fiercely tended,
 Ye little seeds of hate!
I bent above your growing
 Early and noon and late,
Yet are ye drooped and pitiful,--
 I cannot rear ye straight!"

twistd now can accept subcommands equivalent to mktap's;
it will look up the IServiceMaker plugin for the given subcommand
and use it in precisely the same way that mktap does, except that it
immediately applies the service to an Application and starts it.

That is to say,

"twistd web --path ."

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
